### ISSUE TITLE
dugite download unreliable?

### ISSUE THREAD

**Issue by dankegel** [desktop/dugite#195]
Building electron failed for me just now:
```
npm info lifecycle dugite@1.67.0~postinstall: dugite@1.67.0

> dugite@1.67.0 postinstall C:\builds\electry\build\node_modules\dugite
> node ./script/download-git.js

Validation failed. Expected '32364490a9df3baf0ef9dbf56eed14afdd12b6262e887fe324a7246913e439df' but got '14079ad554a215fe707fe64dc3afa041a23d8d846817938f0bacebabe0a61267'
Downloading Git from: https://github.com/desktop/dugite-native/releases/download/v2.17.1-3/dugite-native-v2.17.1-win32.tar.gz
```

When I download by hand, I get the right checksum:
```
$ wget https://github.com/desktop/dugite-native/releases/download/v2.17.1-3/dugite-native-v2.17.1-win32.tar.gz
$ sha256sum dugite-native-v2.17.1-win32.tar.gz 
32364490a9df3baf0ef9dbf56eed14afdd12b6262e887fe324a7246913e439df  dugite-native-v2.17.1-win32.tar.gz
```

And if I just rerun the script, it succeeds:
```
$ node script/download-git.js
Downloading Git from: https://github.com/desktop/dugite-native/releases/download/v2.17.1-3/dugite-native-v2.17.1-win32.tar.gz
```

I haven't built this many times, so the fact that I'm getting one failure suggests that failure
is not uncommon.  Perhaps the download could be retried, and/or some diagnostics added (e.g. expected vs. actual size)...

**Comment by dankegel** [comment]
See also https://www.npmjs.com/package/@jcarras/dugite-win-offline for another example of download problems :-)

**Comment by shiftkey** [comment]
@dankegel this is a side effect of the local cache and the versioning scheme we implemented a while ago.

[`v2.16.2`](https://github.com/desktop/dugite-native/releases/tag/v2.16.2) was the last version to embed the build number in the URL, e.g. `https://github.com/desktop/dugite-native/releases/download/v2.16.2/dugite-native-v2.16.2-macOS-119.tar.gz`.

After that release I started evaluating using CircleCI for packaging for macOS, so we no longer had a consistent build number to rely on when generating assets.

The next release, [`v2.16.2-1`](https://github.com/desktop/dugite-native/releases/tag/v2.16.2-1), dropped the build number, e.g. `https://github.com/desktop/dugite-native/releases/download/v2.16.2-1/dugite-native-v2.16.2-macOS.tar.gz`.

This warning is part of the existing infrastructure setup to handle corrupted downloads or network failure, and because the underlying file isn't different between point releases it sees the old file as not matching and downloads it again.

So I think this behaviour is fine, if a little nuisance. I'm open to improving this, if you have suggestions.

**Comment by dankegel** [comment]
I'm a bit confused.  So you're saying the failure is intentional, but will only happen once per cache?  If that's the case, it's not that bad.

**Comment by shiftkey** [comment]
@dankegel yes, it's not a blocking failure but rather a caching issue and the `npm install` will try again to download the right file - if that doesn't work, it will error.

**Comment by shiftkey** [comment]
> Are you saying it only fails on network error?

Let me try again:

 - installing `dugite` will use a local cache ([see the docs](https://github.com/desktop/dugite/blob/master/docs/environment-variables.md#installation))
 - in your scenario, the new version has the same file name, so it looked at the cache and found the file from the same version
 - the checksum validation failed, so it ignored the cached file and downloaded a fresh file
 - running it again will look at the cache, but then find the right file

**Comment by dankegel** [comment]
I can't quite understand why it downloads the right file and then fails, but whatever.  It did pass the second time around, so I'm not blocked.

(FWIW independent downloads like this are a security risk, you should offer an alternate approach that supports offline builds.  But that's another issue.)

**Comment by dankegel** [comment]
Anyway, thanks for the prompt replies, and let me confess I've shipped cache code that was far worse.

**Comment by dankegel** [comment]
This is biting us again, and this time it's not going away on retries.  Is the workaround to find and delete the cached file?  (Assuming 1.72.0 is still supported...)

```
> dugite@1.72.0 postinstall C:\builds\electry\build\node_modules\dugite
> node ./script/download-git.js

Validation failed. Expected '186cba377cbb7f2bb0d0061ccdd9ddc12c7acc8381a2f56e3c59952a1ecc09a9' but got 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'
Downloading Git from: https://github.com/desktop/dugite-native/releases/download/v2.18.0-4/dugite-native-v2.18.0-windows-x64.tar.gz
Running in verbose mode
git submodule sync --recursive
git submodule update --init --recursive
c:\Python27\python.exe setup.py build
c:\Python27\python.exe setup.py build
npm.cmd install --verbose
None

C:\builds\electry\build>if errorlevel 1 goto :abort 
Traceback (most recent call last):
  File "script/bootstrap.py", line 249, in <module>
    sys.exit(main())
  File "script/bootstrap.py", line 54, in main
    update_node_modules('.')
  File "C:\builds\electry\build\script\lib\util.py", line 289, in update_node_modules
    execute_stdout(args, env)
  File "C:\builds\electry\build\script\lib\util.py", line 184, in execute_stdout
    raise e
subprocess.CalledProcessError: Command '['npm.cmd', 'install', '--verbose']' returned non-zero exit status -1073741819
```

**Comment by dankegel** [comment]
Looks like it.  Should have been a giveaway: that sha256 sum is the one for an empty file :-)

Deleting c:\users\buildbot\AppData\Local\temp\dugite* and rerunning fixed it.

May I vote for a more robust caching behavior?  Removing the downloaded file if its sha256 sum is wrong would probably be a good idea.

**Comment by shiftkey** [comment]
@dankegel apologies for not getting back to this, but I've changed the bits generated from `dugite-native` to embed the source SHA in https://github.com/desktop/dugite-native/pull/125 which is available in `v1.79.0`.

I believe that and #205 should help out here, but let me know if this is still an ongoing issue and we can revisit it.

**Comment by dankegel** [comment]
Thanks.  In the meantime, here's the workaround I'm using to manually load the cache when it get stuck with a size-zero download:
```
  cd /cygdrive/c/users/buildbot/appdata/local/temp/
  rm dugite-native-v2.19.0-windows-x64.tar.gz
  wget https://github.com/desktop/dugite-native/releases/download/v2.19.0-1/dugite-native-v2.19.0-windows-x64.tar.gz
```
